gitbook/graphics-shape.md

```diff
@@ -2,246 +2,234 @@
+window.draw(line, SF::Lines)
 ```
 
 ![A line shape drawn as a primitive](./images/graphics-shape-line-primitive.png)
 
-To learn more about vertices and primitives, you can read the tutorial on [vertex arrays](./graphics-vertex-array.html "'Vertex arrays' tutorial"). 
+To learn more about vertices and primitives, you can read the tutorial on [vertex arrays](graphics-vertex-array.md "'Vertex arrays' tutorial").
 
 ## Custom shape types
 
-You can extend the set of shape classes with your own shape types. To do so, you must derive from [Shape]({{book.api}}/Shape.html) and override two functions: 
+You can extend the set of shape classes with your own shape types. To do so, you must derive from [Shape]({{book.api}}/Shape.html) and override two functions:
 
-  * `getPointCount`: return the number of points in the shape
-  * `getPoint`: return a point of the shape
+  * `point_count`: return the number of points in the shape
+  * `get_point`: return a point of the shape
 
-You must also call the `update()` protected function whenever any point in your shape changes, so that the base class is informed and can update its internal geometry. 
+You must also call the `update()` protected function whenever any point in your shape changes, so that the base class is informed and can update its internal geometry.
 
-Here is a complete example of a custom shape class: EllipseShape. 
+Here is a complete example of a custom shape class: EllipseShape.
 
-```
-class EllipseShape : public sf::Shape
-{
-public :
-
-    explicit EllipseShape(const sf::Vector2f& radius = sf::Vector2f(0, 0)) :
-    m_radius(radius)
-    {
-        update();
-    }
-
-    void setRadius(const sf::Vector2f& radius)
-    {
-        m_radius = radius;
-        update();
-    }
-
-    const sf::Vector2f& getRadius() const
-    {
-        return m_radius;
-    }
-
-    virtual unsigned int getPointCount() const
-    {
-        return 30; // fixed, but could be an attribute of the class if needed
-    }
-
-    virtual sf::Vector2f getPoint(unsigned int index) const
-    {
-        static const float pi = 3.141592654f;
-
-        float angle = index * 2 * pi / getPointCount() - pi / 2;
-        float x = std::cos(angle) * m_radius.x;
-        float y = std::sin(angle) * m_radius.y;
-
-        return sf::Vector2f(m_radius.x + x, m_radius.y + y);
-    }
-
-private :
-
-    sf::Vector2f m_radius;
-};
+```ruby
+# broken https://github.com/manastech/crystal/issues/605
+class EllipseShape < SF::Shape
+    def initialize(radius = SF.vector2(0, 0))
+        @radius = radius
+    end
+
+    def radius=(radius)
+        @radius = radius
+    end
+
+    def radius
+        @radius
+    end
+
+    def point_count
+        30
+    end
+
+    def get_point(index)
+        pi = Math::PI
+
+        angle = index * 2 * pi / point_count - pi / 2;
+        x = Math.cos(angle) * @radius.x
+        y = Math.sin(angle) * @radius.y
+
+        SF.vector2(@radius.x + x, @radius.y + y)
+    end
+end
```

Comment:
Custom shapes currently don't work. Is it related to this? https://github.com/manastech/crystal/issues/605